File: common/src/main/java/com/hartwig/actin/molecular/datamodel/driver/Fusion.java
package com.hartwig.actin.molecular.datamodel.driver;

import org.immutables.value.Value;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Value.Immutable
@Value.Style(passAnnotations = { NotNull.class, Nullable.class })
public abstract class Fusion implements Driver {

    @NotNull
    public abstract String geneStart();

    @NotNull
    public abstract String geneTranscriptStart();

    @NotNull
    public abstract String geneContextStart();

    public abstract int fusedExonUp();

    @NotNull
    public abstract String geneEnd();

    @NotNull
    public abstract String geneTranscriptEnd();

    @NotNull
    public abstract String geneContextEnd();

    public abstract int fusedExonDown();

    @NotNull
    public abstract FusionDriverType driverType();

    @NotNull
    public abstract ProteinEffect proteinEffect();

    @Nullable
    public abstract Boolean isAssociatedWithDrugResistance();

}

File: common/src/main/java/com/hartwig/actin/molecular/datamodel/driver/Fusion.kt
package com.hartwig.actin.molecular.datamodel.driver

import com.hartwig.actin.molecular.datamodel.evidence.ActionableEvidence
import com.hartwig.actin.molecular.sort.driver.FusionComparator

data class Fusion(
    val geneStart: String,
    val geneTranscriptStart: String,
    val fusedExonUp: Int,
    val geneEnd: String,
    val geneTranscriptEnd: String,
    val fusedExonDown: Int,
    val driverType: FusionDriverType,
    val proteinEffect: ProteinEffect,
    val isAssociatedWithDrugResistance: Boolean?,
    override val isReportable: Boolean,
    override val event: String,
    override val driverLikelihood: DriverLikelihood?,
    override val evidence: ActionableEvidence
) : Driver, Comparable<Fusion> {

    override fun compareTo(other: Fusion): Int {
        return FusionComparator().compare(this, other)
    }
}